---
title: "PumilioColoromicsDEAnalyses"
author: "Adam Stuckert"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  html_document: default
  pdf_document: default
---

### Setup

```{r global_options, include=FALSE}
# load packages required for analyses
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE)
library(tximport)
library(DESeq2)
library(dplyr)
library(foreach)
library(data.table)
library(splines)
library(gridExtra)
library(ggplot2)
library("BiocParallel")
register(SnowParam(1))
library(apeglm)
library(tidyverse)
dir.create("resultsRodriguezAssembly")
library(topGO)
library(GO.db)
```

Some functions I wrote for this script.

```{r adam's fancy functions}

# this first function runs a Walds test between specified groups, then returns a dataframe from it.
getDEdf <- function(dds, treatment, group1, group2){
  res <- results(dds, contrast = c(treatment, group1, group2))
  res.df <- setDT(as.data.frame(res), keep.rownames = "Entry_name")
  return(res.df)
}

getLFCdf <- function(dds, treatment, group1, group2){
  ddslfc <- lfcShrink(dds, contrast = c(treatment, group1, group2), type="normal")
  res.df <- setDT(as.data.frame(ddslfc), keep.rownames = "Entry_name")
  return(res.df)
}
#restad26 <- lfcShrink(ddstad2, contrast=c("time.location","GS26_dorsum", #"GS26_venter"), type = "normal")


# this function pulls out particular a priori genes of interest into a new data frame
aprioris <- function(df, annos){
  df <- dplyr::left_join(df, annos, by = "Entry_name")
  df.aprioris <- df %>% filter(Gene_name %in% colors$gene_name )
  return(df.aprioris)
}

# this function extracts all genes below a specified alpha value, prints the number of DE genes to screen, and saves a spreadsheet to file
siggies <- function(df, alpha, csv){
  sigs <- df %>% filter(padj < alpha)
  print(paste0("Number of significant genes: ", length(sigs$padj)))
  write.csv(sigs, csv, row.names = FALSE)
  return(sigs)
}
```

Print R environment information:

```{R session info}
sessionInfo()
```

### Data import

Create sample spreadsheet.

```{r create spreadsheet}
#################### List all samples from expression data ###############
# get the directory/path for each sample in this study
base_dir <- getwd()
# get directories
dirs <- list.dirs(path = "data/Rodriguez_salmon_quantification", full.names = F, recursive = F)
# append full file path to each sample id
files <- file.path(base_dir, "data/Rodriguez_salmon_quantification", dirs, "quant.sf") # files = directory + salmon directory + sample name + quantifictaion file name
names(files) <- dirs
all(file.exists(files)) # do these all actually exist?

#################### Make sample spreadsheet ###############
samples <- as.data.frame(dirs)
samples$sample <- dirs
samples <- samples[,c(2,1)]

# remove extra bits from sample ID
samples$sample <- gsub(pattern = "_S.*$", replacement = "", samples$sample)

# create age class
samples$age.class <- samples$sample
samples$age.class <- gsub(pattern = "AD.*$", replacement = "adult", samples$age.class)
samples$age.class <- gsub(pattern = "GS.*$", replacement = "tadpole", samples$age.class)

# create a bigger age class one for plotting
samples$full.age.class <- samples$sample
samples$full.age.class <- gsub(pattern = "AD.*$", replacement = "adult", samples$full.age.class)
samples$full.age.class <- gsub(pattern = "_.*$", replacement = "", samples$full.age.class)

# create Gosner stage
samples$gosner.stage <- samples$sample
samples$gosner.stage <- gsub(pattern = "AD.*$", replacement = "", samples$gosner.stage)
samples$gosner.stage <- gsub(pattern = "_.*$", replacement = "", samples$gosner.stage)
samples$gosner.stage <- gsub(pattern = "GS", replacement = "", samples$gosner.stage)

# create sampling locality
samples$location <- samples$sample
samples$location <- gsub(pattern = "^.*_D.*$", replacement = "dorsum", samples$location)
samples$location <- gsub(pattern = "^.*_V.*$", replacement = "venter", samples$location)

# save as spreadsheet
dir.create("results")
write.csv(samples, "results/samplespreadsheet.csv", row.names = F)

```

Import expression data.

```{r import expression data}
#################### Import annotation documents ###############
annos <- fread("annotation_documents/pumilio.annotation.Rod.txt", header = FALSE)

# rename columns from annos file
annos <- annos[,c(1,3)]
colnames(annos) <- c("transcript", "Entry")

# import gene names
genes <- fread("annotation_documents/uniprot-reviewed_yes.tab", header = TRUE)

tx2annos <- dplyr::left_join(annos, genes, by = "Entry")
#subset out only tx and protein id columns
tx2gene <- tx2annos[,c(1,3)]
colnames(tx2gene)[2] <- "Entry_name"
tx2gene$Entry_name <- gsub(pattern = "_.*$", replacement = "", tx2gene$Entry_name)


#################### Import expression data ###############
# drop any nas for importing
tx2genenoNA <- drop_na(tx2gene)

# import data via tximport
txi <- tximport(files, type="salmon", tx2gene=tx2genenoNA)
names(txi)

counts <- txi$counts
```

Import color data

```{r import color data}
# Import a priori candidate color gene database
colors <- read.csv("supplemental_documents/color_genes_fromgenomepaper.csv")
colnames(colors)[1] <- "gene_name"
colors$gene_name <- tolower(colors$gene_name)

# make annotation frame to merge
tx2annos2 <- tx2annos[,c(3,6)]
colnames(tx2annos2) <- c("Entry_name","Gene_name")
tx2annos2$Entry_name <- gsub(pattern = "_.*$", replacement = "", tx2annos2$Entry_name)
tx2annos2$Gene_name <- gsub(pattern = "\\s.*$", replacement = "", tx2annos2$Gene_name)
tx2annos2$Gene_name <- tolower(tx2annos2$Gene_name)
tx2annos2$Gene_name <- tolower(tx2annos2$Gene_name)
# remove duplicate rows
tx2annos2 <- tx2annos2[!duplicated(tx2annos2),]
# save annotation file...
write.csv(tx2annos, "resultsRodriguezAssembly/Pumilio_transcript2GeneName.csv", row.names = FALSE)
```

### DESeq data creation

```{r DESEQ creation}
# create DESeq dataset
ddsTC <- DESeqDataSetFromTximport(txi, colData = samples, design = ~ 1)

# filter out low expression contigs
keep <- rowSums(counts(ddsTC)) >= 10
ddsTC <- ddsTC[keep,]


#dds <- DESeq(ddsTC)
#res <- results(dds)
```

### Create a PCA

```{R PCA}
vsd <- vst(ddsTC, blind = FALSE)

# get PC1 and PC2 data
pcaData <- plotPCA(vsd, intgroup = c("full.age.class", "location"), returnData = TRUE)

# get percent variation
percentVar <- round(100 * attr(pcaData, "percentVar"))

# Make a nice naming scheme for plotting
pcaData$age4plot = gsub(pattern = "GS", replacement = "Gosner ", pcaData$full.age.class)
pcaData$age4plot = gsub(pattern = "adult", replacement = "Adult", pcaData$age4plot)
pcaData$age4plot = as.factor(pcaData$age4plot)

# Recode age for order
pcaData$age4plot = factor(pcaData$age4plot, levels = c("Gosner 26", "Gosner 37", "Gosner 42", "Gosner 46", "Adult"))


# pca code
ggplot(pcaData, aes(x = PC1, y = PC2, color = age4plot, shape = location)) +
  stat_ellipse(aes(group = age4plot), type = "t", level = 0.95, size = 1.25, show.legend = FALSE) + geom_point(size = 3, show.legend = TRUE) + 
  scale_color_manual(values = c("cyan3", "magenta2", "gold2", "chartreuse3", "royalblue3")) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() + theme_bw() +
  guides(color = guide_legend(title = "Age", override.aes = list(linetype = 0), order = 1), shape = guide_legend(title = "Location", override.aes = list(linetype = 0), order = 2)) 


# save
dir.create("figures")
#ggsave("figures/MultispeciesPCA.pdf", width = 7.63, height = 7.63, dpi = 400)
ggsave("figures/pumilioPCA.png", width = 8, height = 4.6, dpi = 600)
```

### Adult analyses

Import data

```{r import adult data}
adultsamples <- samples %>% filter(age.class == "adult")
rownames(adultsamples) <- adultsamples$sample
adultfiles <- file.path(base_dir, "data/Rodriguez_salmon_quantification/", adultsamples$dirs, "quant.sf")
names(adultfiles) <- paste0(adultsamples$sample)

#################### Import expression data ###############
# import data via tximport
txiadult <- tximport(adultfiles, type="salmon", tx2gene=tx2genenoNA)
names(txiadult)

adultcounts <- txiadult$counts
write.table(adultcounts, "data/adultcountdata.tsv", sep = "\t")
```

### Adult Walds tests

Run differential expression analysis. What is the effect of color patch in adults?

```{r Walds test adults}
ddsadult <- DESeqDataSetFromTximport(txiadult, colData = adultsamples, design = ~ location)
ddsadult <- DESeq(ddsadult)

# walds test between colors
resultsNames(ddsadult)

res.adult <- getLFCdf(ddsadult, "location", "dorsum", "venter")

# how many are "significant"?
res.adult.siggies <- siggies(res.adult, 0.05, "resultsRodriguezAssembly/adultsignificantgenes.csv")

# get df with color genes
res.adult.colors <- aprioris(res.adult, tx2annos2)

# how many are "significant"?
res.adult.colors.siggies <- siggies(res.adult.colors, 0.05, "resultsRodriguezAssembly/adult_significant_color_genes.csv")
```

### Tadpole tests

Import data

```{r import tadpole data}
tadsamples <- samples %>% filter(age.class == "tadpole")
rownames(tadsamples) <- tadsamples$sample
tadfiles <- file.path(base_dir, "data/Rodriguez_salmon_quantification/", tadsamples$dirs, "quant.sf")
names(tadfiles) <- paste0(tadsamples$sample)

#################### Import expression data ###############
# import data via tximport
txitad <- tximport(tadfiles, type="salmon", tx2gene=tx2genenoNA)
names(txitad)

tadcounts <- txitad$counts
```

### Tadpole LRT test over time

Run differential expression analysis. What is the effect of developmental stage in tadpoles?

```{r LRT throughout development}
ddstad <- DESeqDataSetFromTximport(txitad, colData = tadsamples, design = ~ location + gosner.stage)

#### Note to future Adam: I tried "design = ~ location:gosner.stage + gosner.stage" but it did not produce a model that ran...

# filter out low expression contigs
#keep <- rowSums(counts(ddstad)) >= 10
#ddstad <- ddstad[keep,]

# run LRT test
ddstad <- DESeq(ddstad, test="LRT", reduced = ~ location)
restad <- results(ddstad)
restad$symbol <- mcols(restad)$symbol

# how many are "significant"?
print("Number of significant genes over time in tadpoles: ")
table(restad[,"padj"] < 0.05)


# write to a dataframe and add annotation data
res.tad.df <- setDT(as.data.frame(restad), keep.rownames = "Entry_name")
res.tad.df <- dplyr::left_join(res.tad.df, tx2annos2, by = "Entry_name")

## How many color genes are differentially expressed?
res.tad.colors <- res.tad.df %>% filter(Gene_name %in% colors$gene_name ) 

# print number of significant color genes
print("Number of significant a priori candidate color genes over time in tadpoles: ")
table(res.tad.colors[,"padj"] < 0.05)

# subset out sig color genes
res.tad.colors.sig <- res.tad.colors %>% filter(padj < 0.05)

# save to file
write.csv(res.tad.colors.sig, "tad_significant_color_genes.csv")
```

### Walds tests per time point

I will examine each time point and compare dorsal to venter.

What are the possible comparisons?

```{r tadpole comparisons?}
resultsNames(ddstad)
```

```{r GS26-dorsum to venter comparisons}
# first make a column that is each time point by location
tadsamples$time.location <- paste0(tadsamples$full.age.class, "_", tadsamples$location)

ddstad2 <- DESeqDataSetFromTximport(txitad, colData = tadsamples, design = ~ time.location)
ddstad2 <- DESeq(ddstad2)


# walds test between locations
resultsNames(ddstad2)

#restad26 <- getDEdf(ddstad2, "time.location","GS26_dorsum","GS26_venter")
restad26 <- getLFCdf(ddstad2, "time.location","GS26_dorsum","GS26_venter")
tad26siggies <- siggies(restad26, 0.05, "resultsRodriguezAssembly/tad26significantgenes.csv")
tad26colors <- aprioris(restad26, tx2annos2)
tad26sigcolors <- siggies(tad26colors, 0.05, "resultsRodriguezAssembly/tad26significantcolors.csv")
```

Now run Gosner stage 37.

```{r GS37-dorsum to venter comparisons}
ddstad2$time.location <- factor(ddstad2$time.location, levels = c("GS37_dorsum", "GS37_venter", "GS42_dorsum", "GS42_venter", "GS46_dorsum", "GS46_venter", "GS26_dorsum", "GS26_venter"))
ddstad2 <- DESeq(ddstad2)


restad37 <- getLFCdf(ddstad2, "time.location","GS37_dorsum","GS37_venter")
tad37siggies <- siggies(restad37, 0.05, "resultsRodriguezAssembly/tad37significantgenes.csv")
tad37colors <- aprioris(restad37, tx2annos2)
tad37sigcolors <- siggies(tad37colors, 0.05, "resultsRodriguezAssembly/tad37significantcolors.csv")
```

Now run Gosner stage 42.

```{r GS42-dorsum to venter comparisons}
ddstad2$time.location <- factor(ddstad2$time.location, levels = c("GS42_dorsum", "GS42_venter", "GS46_dorsum", "GS46_venter", "GS26_dorsum", "GS26_venter", "GS37_dorsum", "GS37_venter"))
ddstad2 <- DESeq(ddstad2)


restad42 <- getLFCdf(ddstad2, "time.location","GS42_dorsum","GS42_venter")
tad42siggies <- siggies(restad42, 0.05, "resultsRodriguezAssembly/tad42significantgenes.csv")
tad42colors <- aprioris(restad42, tx2annos2)
tad42sigcolors <- siggies(tad42colors, 0.05, "resultsRodriguezAssembly/tad42significantcolors.csv")
```

Now run Gosner stage 46.

```{r}
ddstad2$time.location <- factor(ddstad2$time.location, levels = c("GS46_dorsum", "GS46_venter", "GS26_dorsum", "GS26_venter", "GS37_dorsum", "GS37_venter", "GS42_dorsum", "GS42_venter"))
ddstad2 <- DESeq(ddstad2)

restad46 <- getLFCdf(ddstad2, "time.location","GS46_dorsum","GS46_venter")
tad46siggies <- siggies(restad46, 0.05, "resultsRodriguezAssembly/tad46significantgenes.csv")
tad46colors <- aprioris(restad46, tx2annos2)
tad46sigcolors <- siggies(tad46colors, 0.05, "resultsRodriguezAssembly/tad46significantcolors.csv")
```

Which genes are differentially expressed between color patches at any time point?

Note: this uses the estimate of log fold change from the "normal" shrinkage from lfcShrink().

```{r overlapping DE gene sets}

# prep
tad26siggiesX <- tad26siggies[,c(1,3)]
colnames(tad26siggiesX)[2] <- "GS26_log2FoldChange"

tad37siggiesX <- tad37siggies[,c(1,3)]
colnames(tad37siggiesX)[2] <- "GS37_log2FoldChange"

tad42siggiesX <- tad42siggies[,c(1,3)]
colnames(tad42siggiesX)[2] <- "GS42_log2FoldChange"


tad46siggiesX <- tad46siggies[,c(1,3)]
colnames(tad46siggiesX)[2] <- "GS46_log2FoldChange"

adultsiggiesX <- res.adult.siggies[,c(1,3)]
colnames(adultsiggiesX)[2] <- "Adult_log2FoldChange"

# merge them all
allDEskinpatches <- list(tad26siggiesX, tad37siggiesX, tad42siggiesX, tad46siggiesX, adultsiggiesX) %>% reduce(full_join, by = "Entry_name")

# add annotation information
allDEskinpatches <- left_join(allDEskinpatches, tx2annos2, by = "Entry_name")

# reorder columns
allDEskinpatches <- allDEskinpatches[,c(7,1:6)]

# save to disk 
write.csv(allDEskinpatches, "resultsRodriguezAssembly/AllDorsumVSVentersignificantgenes.csv")

# print to screen #
print(paste0("Total number of DE genes between dorsum and venter at any time point: ", length(allDEskinpatches$Gene_name)))

# extract just color genes
allDEskinpatchescolorgenes <- allDEskinpatches %>% filter(Gene_name %in% colors$gene_name )

# save to disk 
write.csv(allDEskinpatchescolorgenes, "resultsRodriguezAssembly/AllDorsumVSVentersignificantColorgenes.csv", row.names = FALSE)

# print to screen #
print(paste0("Total number of a priori color genes DE between dorsum and venter at any time point: ", length(allDEskinpatchescolorgenes$Gene_name)))
```

Which genes were differentially expressed between all time points?

```{R DE genes across all time points}
DEalltimepointsskinpatches <- allDEskinpatches %>% drop_na()
DEalltimepointsskinpatches
```

#### Time to make gene figures!

Define aesthetic.

```{r facet aes}
mytheme <- theme_classic() + theme(axis.text=element_text(size=20), 
        axis.title=element_text(size=22,face="bold"), 
        plot.title = element_text(size = 22, hjust = 0.5), strip.text.x = element_text(size = 22),
        panel.border = element_rect(color = "black", fill = NA, size = 1), legend.position = "none")
```

Define plotting functions

```{R adam's fancy plotting functions}
# genes = list of genes
# dds = deseq2 object
# annos = transcript 2 gene name df
extractCounts <- function(genes, dds, annos){
  gene2tx <- annos[annos$Gene_name %in% genes, ]

#### for loop to extract counts
# make empty data frame
countdf <- data.frame()
for (i in 1:nrow(gene2tx)){
  target <- as.character(gene2tx[i,1])
  gene <- as.character(gene2tx[i,2])
  # pull out count data
  tmp <- plotCounts(dds, gene = target, intgroup = "location", returnData = TRUE, xlab="treatment")
  # make rownames a column
  tmp <- setDT(tmp, keep.rownames = TRUE)
  # add gene name to a column
  tmp$Gene_name <- gene
  # add all to a data frame
  countdf <- rbind(countdf, tmp)
  }

  return(countdf)
}

# df = dataframe of genes to plot (output from countdf function)
plotGenes <- function(df, dir){
  dir.create(dir)
  figs <- unique(df$Gene_name)

for (i in 1:length(figs)){
  gene <- figs[i]
  gene_df <- df[df$Gene_name %in% gene, ]
    #times <- tx2annos2[tx2annos2$Gene_name %in% time_genes, ]

ggplot(gene_df,  aes(x = location, y = count)) + 
  ggtitle(gene,  subtitle = "")  + 
  ylab("Normalized Counts") + 
  geom_boxplot(outlier.shape = NA, lwd=.75, aes(fill = location)) + scale_fill_manual(values=c("green4","yellow1")) +
  mytheme
  
ggsave(paste0(dir, "/", gene, ".png"), width = 5.7, height = 4.37)
}
}


extractTadCounts <- function(genes, dds, annos){
  gene2tx <- annos[annos$Gene_name %in% genes, ]

#### for loop to extract counts
# make empty data frame
countdf <- data.frame()
for (i in 1:nrow(gene2tx)){
  target <- as.character(gene2tx[i,1])
  gene <- as.character(gene2tx[i,2])
  # pull out count data
  tmp <- plotCounts(dds, gene = target, intgroup = c("gosner.stage","location"), returnData = TRUE, xlab="treatment")
  # make rownames a column
  tmp <- setDT(tmp, keep.rownames = TRUE)
  # add gene name to a column
  tmp$Gene_name <- gene
  # add all to a data frame
  countdf <- rbind(countdf, tmp)
  }

  return(countdf)
}


plotGenestime <- function(df, dir){
  dir.create(dir)
  figs <- unique(df$Gene_name)

for (i in 1:length(figs)){
  gene <- figs[i]
  gene_df <- df[df$Gene_name %in% gene, ]
    #times <- tx2annos2[tx2annos2$Gene_name %in% time_genes, ]

ggplot(gene_df,  aes(x = gosner.stage, y = count)) + 
  ggtitle(gene,  subtitle = "")  + 
  ylab("Normalized Counts") + 
  geom_boxplot(outlier.shape = NA, lwd=.75, aes(fill = location)) + 
  scale_fill_manual(values=c("green4","yellow1")) +
  facet_grid(. ~ location) +
  mytheme
  
ggsave(paste0(dir, "/", gene, ".png"), width = 11.38, height = 4.37)
}
}

```

```{r try the adults...}
# DE genes for adults are in: res.adult.colors.siggies
adultDEgenes <- unique(res.adult.colors.siggies$Gene_name)
adultgenesdf <- extractCounts(adultDEgenes, ddsadult, tx2annos2)
plotGenes(adultgenesdf, "figures/adultDEcolorgenes")
```

```{r tadpole LRT figures}
tadDEgenes <- unique(res.tad.colors.sig$Gene_name)
tadgenesdf <- extractTadCounts(tadDEgenes, ddstad, tx2annos2)
plotGenestime(tadgenesdf, "figures/tadDEcolorgenes")
```

### 
